docs: Custom rule & plugin tutorial (#17024)
eslint/eslint: docs/src/extend/custom-rule-tutorial.md

---
title: Custom Rule Tutorial
eleventyNavigation:
    key: custom rule tutorial
    parent: create plugins
    title: Custom Rule Tutorial
    order: 1
---

> This tutorial covers how to create a custom rule for ESLint and distribute it with a plugin.
> 
> You can create custom rules to validate if your code meets a certain expectation, and determine what to do if it does not meet that expectation. Plugins package custom rules and other configuration, allowing you to easily share and reuse them in different projects.

Reviewer: ```suggestion
This tutorial covers how to create a custom rule for ESLint and distribute it with a plugin.

You can create custom rules to validate if your code meets a certain expectation, and determine what to do if it does not meet that expectation. Plugins package custom rules and other configurations, allowing you to easily share and reuse them in different projects.
```

To learn more about custom rules and plugins refer to the following documentation:

* [Custom Rules](custom-rules)
* [Plugins](plugins)

## Why Create a Custom Rule?

> Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, to prevent a particular bug from recurring, or to ensure compliance with a style guide.   
> 
> Before creating a custom rule that isn't specific to your company or project, it's worth searching the web to see someone has published a plugin with a custom rule that solves your use case. It's quite possible the rule may already exist.

Reviewer: ```suggestion
Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, prevent a particular bug from recurring or ensure compliance with a style guide.

Before creating a custom rule that isn't specific to your company or project, it's worth searching the web to see if someone has published a plugin with a custom rule that solves your use case. It's quite possible the rule may already exist.
```

## Prerequisites

Before you begin, make sure you have the following installed in your development environment:

* [Node.js](https://nodejs.org/en/download/)
* [npm](https://www.npmjs.com/)

This tutorial also assumes that you have a basic understanding of ESLint and ESLint rules.

## The Custom Rule You Make

> The custom rule that you make in this tutorial requires that all `const` variables named `foo` are assigned the string `"bar"`. The rule is defined in the file `foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`

Reviewer: ```suggestion
The custom rule that you make in this tutorial requires that all `const` variables named `foo` are assigned the string `"bar"`. The rule is defined in the file `foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.
```

You can run the custom rule on a file to report and fix any problems.

For example, say you had the following `foo.js` file:

```javascript
// foo.js

const foo = "baz123";
```

Running ESLint with the rule would fix the file to contain the following:

```javascript
// foo.js

const foo = "bar";
```

## Step 1: Set up Project

> First create a new project for your custom rule. Create a new directory, initiate a new npm project in it, and create a new file for the custom rule:

Reviewer: ```suggestion
First, create a new project for your custom rule. Create a new directory, initiate a new npm project in it, and create a new file for the custom rule:
```

```shell
mkdir eslint-custom-rule-example # create directory
cd eslint-custom-rule-example # enter the directory
npm init -y # init new npm project
touch enforce-foo-bar.js # create file foo-bar.js
```

## Step 2: Stub Out Rule File

> In the `foo-bar.js` file, add some scaffolding for the `foo-bar` custom rule. Also add a `meta` object with some basic information about the rule.

Reviewer: ```suggestion
In the `foo-bar.js` file, add some scaffolding for the `foo-bar` custom rule. Also, add a `meta` object with some basic information about the rule.
```

```javascript
// enforce-foo-bar.js

module.exports = {
    meta: {
        type: "problem",
        docs: {
            description: "Enforce that a variable named `foo` can only be assigned a value of 'bar'.",
        },
        fixable: "code"
    },
    create(context) {
        return {
            // TODO: add callback function(s)
            }
        };
    }
};
```

To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).

## Step 4: Add Rule Visitor Methods

Define the rule's `create` function, which accepts a `context` object and returns an object with a property for each syntax node type you want to handle. In this case, we want to handle `VariableDeclaration` nodes.
You can choose any [ESTree node type](https://github.com/estree/estree) or [selector](selectors).

Inside the `VariableDeclaration` visitor method, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `VariableDeclaration` method.

> If the `const foo` declaration_is assigned to `"bar"` the rule does nothing. If `const foo` **is not** assigned to `"bar"`, then `context.report()` reports an error to ESLint. The error report include information about the error and how to fix it.

Reviewer: ```suggestion
If the `const foo` declaration_is assigned to `"bar"` the rule does nothing. If `const foo` **is not** assigned to `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.
```

                            // Also includes a `fix(fixer)` function that replaces
                            // any values assigned to `const foo` with "bar".
                            context.report({
                                node,
                                message: 'Value other than "bar" assigned to \`const foo\`. Unexpected value: {{ notBar }}',
                                data: {
                                    notBar: node.declaration.init.value
                                },
                                fix(fixer) {
                                    return fixer.replaceText(node.declaration.init, "bar");
                                }
                            });
                        }
                    }
                }
            }
        };
    }
};
```

## Step 5: Set up Testing

With the rule written, you can test it to make sure it's working as expected.

ESLint provides the built-in [`RuleTester`](../integrate/nodejs-api#ruletester) class to test rules. You do not need to use third-party testing libraries to test ESLint rules, but `RuleTester` works seamlessly with tools like Mocha and Jest.

> Next create the file for the tests, `foo-bar.test.js`:

Reviewer: ```suggestion
Next, create the file for the tests, `foo-bar.test.js`:
```

```shell
touch foo-bar.test.js
```

And add a test script to your `package.json` file to run the tests:

```javascript
// package.json
{
    // ...other configuration
    "scripts": {
        "test": "node enforce-foo-bar.test.js"
    },
    // ...other configuration
}
```

## Step 6: Write the Test

To write the test using `RuleTester`, import the class and your custom rule into the `foo-bar.test.js` file.

> The `RuleTester#run()` method tests the rule against valid and invalid test cases. If the rule doesn't match any of the criteria in the tests, the method throws an error.

Reviewer 2: ```suggestion
The `RuleTester#run()` method tests the rule against valid and invalid test cases. If the rule fails to pass any of the test scenarios, this method throws an error.
`RuleTester` requires that at least one valid and one invalid test scenarios be present.
```

  "foo-bar", // rule name
  fooBarRule, // rule code
  { // checks
    // 'valid' checks cases that should pass
    valid: [{
      code: "const foo = 'bar';",
    }],
    // 'invalid' checks cases that should not pass
    invalid: [{
      code:"const foo = 'baz';",
      output: 'const foo = "bar";',
      errors: 1,
    }],
  }
);

console.log("All tests passed!");
```

## Step 7: Bundle the Custom Rule in a Plugin

Now that you've written the custom rule and validated that it works, you can include it in a plugin. Using a plugin, you can share the rule in an npm package to use in other projects.

Create the file for the plugin:

```shell
touch eslint-plugin-example.js
```

And now write the plugin code. Plugins are just exported JavaScript objects. To include a rule in a plugin, include it in the plugin's `rules` object, which contains key-value pairs of rule names and their source code.

To learn more about creating plugins, refer to [Create Plugins](plugins).

```javascript
// eslint-plugin-example.js

const fooBarRule = require("./enforce-foo-bar");
const plugin = { rules: { "foo-bar": fooBarRule } };
module.exports = plugin;
```

## Step 8: Publish the Plugin

> To publish a plugin containing a rule to npm, you need configure the `package.json`. Add the following in the corresponding fields:

Reviewer: ```suggestion
To publish a plugin containing a rule to npm, you need to configure the `package.json`. Add the following in the corresponding fields:
```

1. `"name"`: A unique name for the package. No other package on npm can have the same name.
1. `"main"`: The relative path to the plugin file. Following this example, the path is `"eslint-plugin-example.js"`.
1. `"description"`: A description of the package that's viewable on npm.

> 1. `"peerDependencies"`: Add `"eslint": ">=8.0.0"` as a peer dependency. Any version greater than or equal to that is necessary to use the plugin. By declaring `eslint` as a peer dependency, it requires that users add the package to the project separately from the plugin.

Reviewer: ```suggestion
1. `"peerDependencies"`: Add `"eslint": ">=8.0.0"` as a peer dependency. Any version greater than or equal to that is necessary to use the plugin. Declaring `eslint` as a peer dependency requires that users add the package to the project separately from the plugin.
```

  "peerDependencies": {
    "eslint": ">=8.0.0"
  },
  // Add these standard keywords to make plugin easy to find!
  "keywords": [
    "eslint",
    "eslintplugin",
    "eslint-plugin"
  ],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "eslint": "^8.36.0",
    "np": "^7.6.3"
  }
}
```

To publish the package, run `npm publish` and follow the CLI prompts.

You should see the package live on npm!

## Step 9: Use the Published Custom Rule

To use the package now that it's published, run the following command in your project to download your published package:

```shell
npm install --save-dev eslint-plugin-example # Add your package name here
```

Then create an ESLint configuration for your project using a [flat configuration file](../use/configure/configuration-files-new), `eslint.config.js`.

Add the following code to `eslint.config.js`:

```javascript
// eslint.config.js
"use strict";

// Import the ESLint plugin
const eslintPluginExample = require("./eslint-plugin-example");

module.exports = [
    {
        files: ["**/*.js"],
        languageOptions: {
            sourceType: "commonjs",
            ecmaVersion: "latest",
        },

>         // Using the eslint-plugin-foo-bar plugin downloaded from npm

Author: thanks for the catch! yes, it should be `eslint-plugin-example`.

>         // Using the eslint-plugin-foo-bar plugin downloaded from npm

Author: ```suggestion
        // Using the eslint-plugin-example plugin downloaded from npm
```

>         plugins: {"example": eslintPluginExample},

Author: @moniuch we are actually defining the name as 'example' here (not a shorthand), as this is following the new config format.

npx eslint example.js
```

This produces the following output in the terminal:

```text
/<path-to-directory>/eslint-custom-rule-example/example.js
  14:3  error  Value other than "bar" assigned to `const foo`. Unexpected value: baz  foo-bar/foo-bar

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

As you can see in the above message, you can actually fix the issue with the `--fix` flag, correcting the variable assignment to be `"bar"`.

Run ESLint again with the `--fix` flag:

```shell
npx eslint example.js --fix
```

There is no error output in the terminal when you run this, but you can see the fix applied in `example.js`. You should see the following:

```javascript
// example.js

// ... rest of file

function incorrectFoo(){
  const foo = "bar"; // Fixed!
}
```

## Summary

In this tutorial, you've learned how to create a custom rule and plugin for ESLint. You've made a custom rule that requires all `const` variables named `foo` to be assigned the string `"bar"` and suggests replacing any other value assigned to `const foo` with `"bar"`. You've also added the rule to a plugin, and published the plugin on npm.

> Through doing this, you've learned the following practices which you can apply to creating other custom rules and plugins:

Reviewer: ```suggestion
Through doing this, you've learned the following practices which you can apply to create other custom rules and plugins:
```

        },
        fixable: "code"
    },
    create(context) {
        return {
            // TODO: add callback function(s)
        };
    }
};
```

To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).

## Step 4: Add Rule Visitor Methods

Define the rule's `create` function, which accepts a `context` object and returns an object with a property for each syntax node type you want to handle. In this case, you want to handle `VariableDeclaration` nodes.
You can choose any [ESTree node type](https://github.com/estree/estree) or [selector](selectors).

Inside the `VariableDeclaration` visitor method, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `VariableDeclaration` method.

If the `const foo` declaration is assigned a value of `"bar"`, then the rule does nothing. If `const foo` **is not** assigned a value of `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.

```javascript
// enforce-foo-bar.js

> module.exports = {

Author: ```suggestion
{% raw %}
module.exports = {
```

                        // Check if the value of the variable is "bar"
                        if (node.declaration[0].init.value !== "bar") {
                            // Report error to ESLint. Error message uses
                            // a message placeholder to include the incorrect value
                            // in the error message.
                            // Also includes a `fix(fixer)` function that replaces
                            // any values assigned to `const foo` with "bar".
                            context.report({
                                node,
                                message: 'Value other than "bar" assigned to \`const foo\`. Unexpected value: {{ notBar }}',
                                data: {
                                    notBar: node.declaration.init.value
                                },
                                fix(fixer) {
                                    return fixer.replaceText(node.declarations[0].init, '"bar"');
                                }
                            });
                        }
                    }
                }
            }
        };
    }
};

> ```

Author: ```suggestion
{% endraw %}
```
```

                        // Check if the value of the variable is "bar"
                        if (node.declaration[0].init.value !== "bar") {
                            // Report error to ESLint. Error message uses
                            // a message placeholder to include the incorrect value
                            // in the error message.
                            // Also includes a `fix(fixer)` function that replaces
                            // any values assigned to `const foo` with "bar".
                            context.report({
                                node,
                                message: 'Value other than "bar" assigned to \`const foo\`. Unexpected value: {{ notBar }}',
                                data: {
                                    notBar: node.declaration.init.value
                                },
                                fix(fixer) {
                                    return fixer.replaceText(node.declarations[0].init, '"bar"');
                                }
                            });
                        }
                    }
                }
            }
        };
    }
};

> {% endraw %}

Author: ```suggestion
{% endraw %}
```
```

1. `"keywords"`: Include the standard keywords `["eslint", "eslintplugin", "eslint-plugin"]` to make the package easy to find. You can add any other keywords that might be relevant to your plugin as well.

A complete annotated example of what a plugin's `package.json` file should look like:

```javascript
// package.json
{
  // Name npm package.
  // Add your own package name. eslint-plugin-example is taken!
  "name": "eslint-plugin-example",
  "version": "1.2.0",
  "description": "ESLint plugin for foo-bar rule.",
  "main": "eslint-plugin-example.js", // plugin entry point
  "scripts": {
    "test": "node enforce-foo-bar.test.js"
  },
  // Add eslint>=8.0.0 as a peer dependency.
  "peerDependencies": {
    "eslint": ">=8.0.0"
  },
  // Add these standard keywords to make plugin easy to find!
  "keywords": [
    "eslint",
    "eslintplugin",
    "eslint-plugin"
  ],
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "eslint": "^8.36.0",

>     "np": "^7.6.3"

Author: ```suggestion
```

Inside the `VariableDeclaration` visitor method, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `VariableDeclaration` method.

If the `const foo` declaration is assigned a value of `"bar"`, then the rule does nothing. If `const foo` **is not** assigned a value of `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.

```javascript
// enforce-foo-bar.js
module.exports = {
    meta: {
        type: "problem",
        docs: {
            description: "Enforce that a variable named `foo` can only be assigned a value of 'bar'.",
        },
         fixable: "code"
    },
    create(context) {
        return {
            // Performs action in the function on every variable declaration
            VariableDeclaration(node) {
                // Check if a `const` variable declaration
                if(node.kind === "const") {
                    // Check if variable name is `foo`

>                     if(node.declarations[0].id.name === "foo") {

Reviewer 3: This checks only the first VariableDeclarator. Should the rule catch e.g. `foo = "baz"` in `const a = 1, foo = "baz";`?

Author: for the sake of simplicity, i think it's fine as is. but if you feel otherwise, we can refactor.

                        // Check if the value of the variable is "bar"
                        if (node.declaration[0].init.value !== "bar") {
                            // Report error to ESLint. Error message uses
                            // a message placeholder to include the incorrect value
                            // in the error message.
                            // Also includes a `fix(fixer)` function that replaces
                            // any values assigned to `const foo` with "bar".
                            context.report({
                                node,
                                message: 'Value other than "bar" assigned to \`const foo\`. Unexpected value: {{ notBar }}',
                                data: {

>                                     notBar: node.declaration.init.value

Reviewer 3: ```suggestion
                                    notBar: node.declarations[0].init.value
```

Also, it looks like this assumes that `node.declarations[0].init` is a Literal node. Should the rule catch only literals (in which case we should update the rule code) or any nodes (in which case we should update this line)?

Author: to handle other types of nodes, what would the code update be here? and what other types of non-literal nodes would make sense to test?

Author: toward the top of the text, updating to be '"assigned to the string **literal** `"bar"`':

```md
## The Custom Rule

The custom rule  in this tutorial requires that all `const` variables named `foo` are assigned the string literal `"bar"`. The rule is defined in the file `enforce-foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.
```

i'd rather not get to deep into this topic since i think it'd be a distraction from the beginner-oriented tutorial

                                },
                                fix(fixer) {

>                                     return fixer.replaceText(node.declaration.init, "bar");

Reviewer 3: ```suggestion
                                    return fixer.replaceText(node.declarations[0].init, '"bar"');
```

Next, create the file for the tests, `enforce-foo-bar.test.js`:

```shell
touch enforce-foo-bar.test.js
```

And add a test script to your `package.json` file to run the tests:

```javascript
// package.json
{
    // ...other configuration
    "scripts": {
        "test": "node enforce-foo-bar.test.js"
    },
    // ...other configuration
}
```

## Step 6: Write the Test

To write the test using `RuleTester`, import the class and your custom rule into the `enforce-foo-bar.test.js` file.

The `RuleTester#run()` method tests the rule against valid and invalid test cases. If the rule fails to pass any of the test scenarios, this method throws an error.
`RuleTester` requires that at least one valid and one invalid test scenarios be present.

```javascript
// enforce-foo-bar.test.js
const {RuleTester} = require("eslint");
const fooBarRule = require("./enforce-foo-bar");

const ruleTester = new RuleTester({
  // Must use at least ecmaVersion 2015 because

>   // that's when `const` variable introduced.

Reviewer 3: ```suggestion
  // that's when `const` variables were introduced.
```

Maybe this?

  // that's when `const` variables were introduced.
  parserOptions: { ecmaVersion: 2015 }
});

// Throws error if the tests in ruleTester.run() do not pass
ruleTester.run(

>   "foo-bar", // rule name

Author: ```suggestion
  "enforce-foo-bar", // rule name
```

  { // checks
    // 'valid' checks cases that should pass
    valid: [{
      code: "const foo = 'bar';",
    }],
    // 'invalid' checks cases that should not pass
    invalid: [{

>       code:"const foo = 'baz';",

Reviewer 3: ```suggestion
      code: "const foo = 'baz';",
```

      output: 'const foo = "bar";',
      errors: 1,
    }],
  }
);

console.log("All tests passed!");
```

## Step 7: Bundle the Custom Rule in a Plugin

Now that you've written the custom rule and validated that it works, you can include it in a plugin. Using a plugin, you can share the rule in an npm package to use in other projects.

Create the file for the plugin:

```shell
touch eslint-plugin-example.js
```

And now write the plugin code. Plugins are just exported JavaScript objects. To include a rule in a plugin, include it in the plugin's `rules` object, which contains key-value pairs of rule names and their source code.

To learn more about creating plugins, refer to [Create Plugins](plugins).

```javascript
// eslint-plugin-example.js

const fooBarRule = require("./enforce-foo-bar");

> const plugin = { rules: { "foo-bar": fooBarRule } };

Reviewer 3: I think the rule name should be the same as the file name, either both `enforce-foo-bar` or both `foo-bar`.

Author: done

> const plugin = { rules: { "foo-bar": fooBarRule } };

Author: ```suggestion
const plugin = { rules: { "enforce-foo-bar": fooBarRule } };
```

> const plugin = { rules: { "foo-bar": fooBarRule } };

Author: ```suggestion
const plugin = { rules: { "enforce-foo-bar": fooBarRule } };
```

```

## Step 8: Use the Plugin Locally

You can use a locally defined plugin to execute the custom rule in your project. To use a local plugin, specify the path to the plugin in the `plugins` property of your ESLint configuration file.

You might want to use a locally defined plugin in one of the following scenarios:

* You want to test the plugin before publishing it to npm.
* You want to use a plugin, but do not want to publish it to npm.

Before you can add the plugin to the project, create an ESLint configuration for your project using a [flat configuration file](../use/configure/configuration-files-new), `eslint.config.js`:

```shell
touch eslint.config.js
```

Then, add the following code to `eslint.config.js`:

```javascript
// eslint.config.js
"use strict";

// Import the ESLint plugin locally
const eslintPluginExample = require("./eslint-plugin-example");

module.exports = [
    {
        files: ["**/*.js"],
        languageOptions: {
            sourceType: "commonjs",
            ecmaVersion: "latest",
        },

>         // Using the eslint-plugin-example plugin downloaded from npm

Reviewer 3: This configuration uses local file `./eslint-plugin-example.js`, not a plugin downloaded from npm.

Author: ```suggestion
        // Using the eslint-plugin-example plugin defined locally
```

Before creating a custom rule that isn't specific to your company or project, it's worth searching the web to see if someone has published a plugin with a custom rule that solves your use case. It's quite possible the rule may already exist.

## Prerequisites

Before you begin, make sure you have the following installed in your development environment:

* [Node.js](https://nodejs.org/en/download/)
* [npm](https://www.npmjs.com/)

This tutorial also assumes that you have a basic understanding of ESLint and ESLint rules.

## The Custom Rule You Make

The custom rule that you make in this tutorial requires that all `const` variables named `foo` are assigned the string `"bar"`. The rule is defined in the file `foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.

You can run the custom rule on a file to report and fix any problems.

For example, say you had the following `foo.js` file:

```javascript
// foo.js

const foo = "baz123";
```

> Running ESLint with the rule would fix the file to contain the following:

Reviewer 4: ```suggestion
Running ESLint with the rule would flag `"baz123"` as an incorrect value for variable `foo`. If ESLint is running in autofix mode, then ESLint would fix the file to contain the following:
```

```javascript
// foo.js

const foo = "bar";
```

> ## Step 1: Set up Project

Reviewer 4: Let's make it a bit more personal
```suggestion
## Step 1: Set up Your Project
```

First, create a new project for your custom rule. Create a new directory, initiate a new npm project in it, and create a new file for the custom rule:

```shell
mkdir eslint-custom-rule-example # create directory
cd eslint-custom-rule-example # enter the directory
npm init -y # init new npm project
touch enforce-foo-bar.js # create file foo-bar.js
```

> ## Step 2: Stub Out Rule File

Reviewer 4: ```suggestion
## Step 2: Stub Out the Rule File
```

> In the `foo-bar.js` file, add some scaffolding for the `foo-bar` custom rule. Also, add a `meta` object with some basic information about the rule.

Reviewer 4: ```suggestion
In the `enforce-foo-bar.js` file, add some scaffolding for the `enforce-foo-bar` custom rule. Also, add a `meta` object with some basic information about the rule.
```

> All ESLint rules are objects that follow this structure.

Reviewer 4: Again, not sure if this sentence adds anything here or actually confuses the progression.

```javascript
// enforce-foo-bar.js

module.exports = {

>   meta: {
>      // TODO: add metadata
>     },
>     create(context) {
>         return {
>             // TODO: add callback function(s)
>         };
>     }
> };

Reviewer 4: Indentation issues

```suggestion
    meta: {
       // TODO: add metadata
    },
    create(context) {
        return {
            // TODO: add callback function(s)
        };
    }
};
```

```

## Step 3: Add Rule Metadata

Before writing the rule, add some metadata to the rule object. ESLint uses this information when running the rule.

> Start by exporting a module with a `meta` object containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "fix," the description is "Can only assign 'bar' to const foo.", and the rule is fixable by modifying the code.

Reviewer 4: Looks like the code and the description have fallen out of sync. Please double check throughout.

```suggestion
Start by exporting an object with a `meta` property containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "problem," the description is "Enforce that a variable named `foo` can only be assigned a value of 'bar'.", and the rule is fixable by modifying the code.
```

```javascript
// enforce-foo-bar.js

module.exports = {
    meta: {
        type: "problem",
        docs: {
            description: "Enforce that a variable named `foo` can only be assigned a value of 'bar'.",
        },
        fixable: "code"
    },
    create(context) {
        return {
            // TODO: add callback function(s)
            }
        };
    }
};
```

To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).

## Step 4: Add Rule Visitor Methods

> Define the rule's `create` function, which accepts a `context` object and returns an object with a property for each syntax node type you want to handle. In this case, we want to handle `VariableDeclaration` nodes.

Reviewer 4: Previously we used "you" in this tutorial, but here we use "we". Let's stick with "you"?

Author: ```suggestion
Define the rule's `create` function, which accepts a `context` object and returns an object with a property for each syntax node type you want to handle. In this case, you want to handle `VariableDeclaration` nodes.
```

You can choose any [ESTree node type](https://github.com/estree/estree) or [selector](selectors).

Inside the `VariableDeclaration` visitor method, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `VariableDeclaration` method.

> If the `const foo` declaration_is assigned to `"bar"` the rule does nothing. If `const foo` **is not** assigned to `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.

Reviewer 4: ```suggestion
If the `const foo` declaration is assigned a value of `"bar"` then the rule does nothing. If `const foo` **is not** assigned a value of `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.
```

Author: ```suggestion
If the `const foo` declaration is assigned a value of `"bar"`, then the rule does nothing. If `const foo` **is not** assigned a value of `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.
```

```javascript
// eslint.config.js
"use strict";

// Import the plugin downloaded from npm
const eslintPluginExample = require("eslint-plugin-example");

// ... rest of configuration
```

Now you're ready to test the custom rule.

Run ESLint on the `example.js` file you created in step 8, now with the downloaded plugin:

```shell
npx eslint example.js
```

This produces the following output in the terminal:

```text
/<path-to-directory>/eslint-custom-rule-example/example.js
  14:3  error  Value other than "bar" assigned to `const foo`. Unexpected value: baz  foo-bar/foo-bar

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

As you can see in the above message, you can actually fix the issue with the `--fix` flag, correcting the variable assignment to be `"bar"`.

Run ESLint again with the `--fix` flag:

```shell
npx eslint example.js --fix
```

There is no error output in the terminal when you run this, but you can see the fix applied in `example.js`. You should see the following:

```javascript
// example.js

// ... rest of file

function incorrectFoo(){
  const foo = "bar"; // Fixed!
}
```

## Summary

> In this tutorial, you've learned how to create a custom rule and plugin for ESLint. You've made a custom rule that requires all `const` variables named `foo` to be assigned the string `"bar"` and suggests replacing any other value assigned to `const foo` with `"bar"`. You've also added the rule to a plugin, and published the plugin on npm.

Reviewer 4: ```suggestion
In this tutorial, you've made a custom rule that requires all `const` variables named `foo` to be assigned the string `"bar"` and suggests replacing any other value assigned to `const foo` with `"bar"`. You've also added the rule to a plugin, and published the plugin on npm.
```

                                data: {
                                    notBar: node.init.value
                                },
                                fix(fixer) {
                                    return fixer.replaceText(node.init, '"bar"');
                                }
                            });
                        }
                    }
                }
            }
        };
    }
};
{% endraw %}
```

## Step 5: Set up Testing

With the rule written, you can test it to make sure it's working as expected.

ESLint provides the built-in [`RuleTester`](../integrate/nodejs-api#ruletester) class to test rules. You do not need to use third-party testing libraries to test ESLint rules, but `RuleTester` works seamlessly with tools like Mocha and Jest.

Next, create the file for the tests, `enforce-foo-bar.test.js`:

```shell
touch enforce-foo-bar.test.js
```

And add a test script to your `package.json` file to run the tests:

```javascript
// package.json
{
    // ...other configuration
    "scripts": {
        "test": "node enforce-foo-bar.test.js"
    },
    // ...other configuration
}
```

> 

Author: ```suggestion

Also install `eslint` as a development dependency:

```shell
npm install eslint --save-dev
```
```

}
```

## Step 6: Write the Test

To write the test using `RuleTester`, import the class and your custom rule into the `enforce-foo-bar.test.js` file.

The `RuleTester#run()` method tests the rule against valid and invalid test cases. If the rule fails to pass any of the test scenarios, this method throws an error.
`RuleTester` requires that at least one valid and one invalid test scenarios be present.

```javascript
// enforce-foo-bar.test.js
const {RuleTester} = require("eslint");
const fooBarRule = require("./enforce-foo-bar");

const ruleTester = new RuleTester({
  // Must use at least ecmaVersion 2015 because
  // that's when `const` variables were introduced.
  parserOptions: { ecmaVersion: 2015 }
});

// Throws error if the tests in ruleTester.run() do not pass
ruleTester.run(
  "enforce-foo-bar", // rule name
  fooBarRule, // rule code
  { // checks
    // 'valid' checks cases that should pass
    valid: [{
      code: "const foo = 'bar';",
    }],
    // 'invalid' checks cases that should not pass
    invalid: [{
      code: "const foo = 'baz';",
      output: 'const foo = "bar";',
      errors: 1,
    }],
  }
);

console.log("All tests passed!");
```

> 

Reviewer 3: We described how to add a test file and a test script. It would be good to also describe how to run the tests (`npm test`).

Also, for this to work there should be a step where `eslint` is locally installed (`npm install --save-dev eslint`)

Author: ````suggestion

Run the test with the following command:

```shell
npm test
```

You should see the following in your console:

```shell
All tests passed!
```

````

1. `"keywords"`: Include the standard keywords `["eslint", "eslintplugin", "eslint-plugin"]` to make the package easy to find. You can add any other keywords that might be relevant to your plugin as well.

A complete annotated example of what a plugin's `package.json` file should look like:

```javascript
// package.json
{
  // Name npm package.
  // Add your own package name. eslint-plugin-example is taken!
  "name": "eslint-plugin-example",
  "version": "1.2.0",
  "description": "ESLint plugin for foo-bar rule.",
  "main": "eslint-plugin-example.js", // plugin entry point
  "scripts": {
    "test": "node enforce-foo-bar.test.js"
  },
  // Add eslint>=8.0.0 as a peer dependency.
  "peerDependencies": {
    "eslint": ">=8.0.0"
  },
  // Add these standard keywords to make plugin easy to find!
  "keywords": [
    "eslint",
    "eslintplugin",
    "eslint-plugin"
  ],
  "author": "",
  "license": "ISC",
  "devDependencies": {

>     "eslint": "^8.36.0",

Reviewer 3: ```suggestion
    "eslint": "^8.36.0"
```

Trailing commas are not valid in JSON.

  }
}
```

To publish the package, run `npm publish` and follow the CLI prompts.

You should see the package live on npm!

## Step 10: Use the Published Custom Rule

Next, you can use the published plugin.

Run the following command in your project to download the package:

```shell
npm install --save-dev eslint-plugin-example # Add your package name here
```

Update the `eslint.config.js` to use the packaged version of the plugin:

```javascript
// eslint.config.js
"use strict";

// Import the plugin downloaded from npm
const eslintPluginExample = require("eslint-plugin-example");

// ... rest of configuration
```

Now you're ready to test the custom rule.

Run ESLint on the `example.js` file you created in step 8, now with the downloaded plugin:

```shell
npx eslint example.js
```

This produces the following output in the terminal:

```text
/<path-to-directory>/eslint-custom-rule-example/example.js

>   14:3  error  Value other than "bar" assigned to `const foo`. Unexpected value: baz  foo-bar/foo-bar

Reviewer 3: ```suggestion
  9:11  error  Value other than "bar" assigned to `const foo`. Unexpected value: baz  example/enforce-foo-bar
```

---
title: Custom Rule Tutorial
eleventyNavigation:
    key: custom rule tutorial
    parent: create plugins
    title: Custom Rule Tutorial
    order: 1
---
This tutorial covers how to create a custom rule for ESLint and distribute it with a plugin.

You can create custom rules to validate if your code meets a certain expectation, and determine what to do if it does not meet that expectation. Plugins package custom rules and other configuration, allowing you to easily share and reuse them in different projects.

To learn more about custom rules and plugins refer to the following documentation:

* [Custom Rules](custom-rules)
* [Plugins](plugins)

## Why Create a Custom Rule?

> Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, prevent a particular bug from recurring or ensure compliance with a style guide.

Reviewer: ```suggestion
Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, prevent a particular bug from recurring, or ensure compliance with a style guide.
```

```

## Step 7: Bundle the Custom Rule in a Plugin

Now that you've written the custom rule and validated that it works, you can include it in a plugin. Using a plugin, you can share the rule in an npm package to use in other projects.

Create the file for the plugin:

```shell
touch eslint-plugin-example.js
```

And now write the plugin code. Plugins are just exported JavaScript objects. To include a rule in a plugin, include it in the plugin's `rules` object, which contains key-value pairs of rule names and their source code.

To learn more about creating plugins, refer to [Create Plugins](plugins).

```javascript
// eslint-plugin-example.js

const fooBarRule = require("./enforce-foo-bar");
const plugin = { rules: { "foo-bar": fooBarRule } };
module.exports = plugin;
```

## Step 8: Use the Plugin Locally

You can use a locally defined plugin in your project by specifying the path to the plugin in the `plugins` property of your ESLint configuration file.

You might want to use a locally defined plugin in one of the following scenarios:

* You want to test the plugin before publishing it to npm.
* You want to use a plugin, but do not want to publish it to npm.

> First create the ESLint configuration file to include the plugin.

Reviewer: ```suggestion
First, create the ESLint configuration file to include the plugin.
```

create an ESLint configuration for your project using a [flat configuration file](../use/configure/configuration-files-new), `eslint.config.js`.

Create the ESLint configuration file `eslint.config.js`:

```shell
touch eslint.config.js
```

Then add the following code to `eslint.config.js`:

> 
> ```javascript
> ```

Reviewer: I believe this block is redundant.
<img width="783" alt="Screenshot 2023-04-29 at 7 39 47 AM" src="https://user-images.githubusercontent.com/46647141/235278897-86b8b0f6-b118-4a2d-9511-f264812c7590.png">

```suggestion
```

  14:3  error  Value other than "bar" assigned to `const foo`. Unexpected value: baz  foo-bar/foo-bar

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

As you can see in the above message, you can actually fix the issue with the `--fix` flag, correcting the variable assignment to be `"bar"`.

Run ESLint again with the `--fix` flag:

```shell
npx eslint example.js --fix
```

There is no error output in the terminal when you run this, but you can see the fix applied in `example.js`. You should see the following:

```javascript
// example.js

// ... rest of file

function incorrectFoo(){
  const foo = "bar"; // Fixed!
}
```

## Summary

In this tutorial, you've learned how to create a custom rule and plugin for ESLint. You've made a custom rule that requires all `const` variables named `foo` to be assigned the string `"bar"` and suggests replacing any other value assigned to `const foo` with `"bar"`. You've also added the rule to a plugin, and published the plugin on npm.

Through doing this, you've learned the following practices which you can apply to create other custom rules and plugins:

> 1. Creating an ESLint custom rule

Reviewer: How about this?

```suggestion
1. Creating a custom ESLint rule
```

To learn more about custom rules and plugins refer to the following documentation:

* [Custom Rules](custom-rules)
* [Plugins](plugins)

## Why Create a Custom Rule?

Create a custom rule if the ESLint [built-in rules](../rules/) and community-published custom rules do not meet your needs. You might create a custom rule to enforce a best practice for your company or project, prevent a particular bug from recurring, or ensure compliance with a style guide.

Before creating a custom rule that isn't specific to your company or project, it's worth searching the web to see if someone has published a plugin with a custom rule that solves your use case. It's quite possible the rule may already exist.

## Prerequisites

Before you begin, make sure you have the following installed in your development environment:

* [Node.js](https://nodejs.org/en/download/)
* [npm](https://www.npmjs.com/)

This tutorial also assumes that you have a basic understanding of ESLint and ESLint rules.

## The Custom Rule

> The custom rule  in this tutorial requires that all `const` variables named `foo` are assigned the string literal `"bar"`. The rule is defined in the file `enforce-foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.

Reviewer: ```suggestion
The custom rule in this tutorial requires that all `const` variables named `foo` are assigned the string literal `"bar"`. The rule is defined in the file `enforce-foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.
```

```javascript
// enforce-foo-bar.js

module.exports = {
    meta: {
       // TODO: add metadata
    },
    create(context) {
        return {
            // TODO: add callback function(s)
        };
    }
};
```

## Step 3: Add Rule Metadata

Before writing the rule, add some metadata to the rule object. ESLint uses this information when running the rule.

Start by exporting an object with a `meta` property containing the rule's metadata, such as the rule type, documentation, and fixability. In this case, the rule type is "problem," the description is "Enforce that a variable named `foo` can only be assigned a value of 'bar'.", and the rule is fixable by modifying the code.

```javascript
// enforce-foo-bar.js

module.exports = {
    meta: {
        type: "problem",
        docs: {
            description: "Enforce that a variable named `foo` can only be assigned a value of 'bar'.",
        },
        fixable: "code",
        schema: []
    },
    create(context) {
        return {
            // TODO: add callback function(s)
        };
    }
};
```

> To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).

Reviewer 3: The current version of custom-rules doesn't have a Rule Structure header, so this fails the Verify Files-Validate internal links check. We could change this link for now and update it later when we update the custom-rules document.

Reviewer 3: Actually, it might be better to just wait for the new version of custom-rules to be merged into main branch, and then merge this as-is.

Reviewer 3: Verify Files-Validate internal links check is okay now.

            }
        };
    }
};
{% endraw %}
```

## Step 5: Set up Testing

With the rule written, you can test it to make sure it's working as expected.

ESLint provides the built-in [`RuleTester`](../integrate/nodejs-api#ruletester) class to test rules. You do not need to use third-party testing libraries to test ESLint rules, but `RuleTester` works seamlessly with tools like Mocha and Jest.

Next, create the file for the tests, `enforce-foo-bar.test.js`:

```shell
touch enforce-foo-bar.test.js
```

You will use the `eslint` package in the test file. Install it as a development dependency:

```shell
npm install eslint --save-dev
```

And add a test script to your `package.json` file to run the tests:

```javascript
// package.json
{
    // ...other configuration
    "scripts": {
        "test": "node enforce-foo-bar.test.js"
    },
    // ...other configuration
}
```

## Step 6: Write the Test

To write the test using `RuleTester`, import the class and your custom rule into the `enforce-foo-bar.test.js` file.

The `RuleTester#run()` method tests the rule against valid and invalid test cases. If the rule fails to pass any of the test scenarios, this method throws an error.

> `RuleTester` requires that at least one valid and one invalid test scenarios be present.

Reviewer: ```suggestion
`RuleTester` requires that at least one valid and one invalid test scenario be present.
```
